File: src/purra/delegation/policy.py
"""Host-configured limits for model-defined delegation."""

from __future__ import annotations

from dataclasses import dataclass

from purra.contracts import DelegationContextMode, ToolExecutionMode
from purra.normalization import positive_int


@dataclass(frozen=True, slots=True)
class DelegationPolicy:
    """Bounds model-defined Agents without granting them new authority."""

    max_agents_per_call: int = 3
    max_parallel: int = 3
    max_agent_name_chars: int = 64
    max_title_chars: int = 120
    max_instruction_chars: int = 4_000
    max_objective_chars: int = 4_000
    max_depth: int = 3
    max_agents_per_root: int = 16
    allow_recursive_delegation: bool = False

    def __post_init__(self) -> None:
        for name in (
            "max_agents_per_call",
            "max_parallel",
            "max_agent_name_chars",
            "max_title_chars",
            "max_instruction_chars",
            "max_objective_chars",
            "max_depth",
            "max_agents_per_root",
        ):
            object.__setattr__(self, name, positive_int(getattr(self, name), name))
        if self.max_parallel > self.max_agents_per_call:
            raise ValueError("max_parallel cannot exceed max_agents_per_call")
        if not isinstance(self.allow_recursive_delegation, bool):
            raise TypeError("allow_recursive_delegation must be boolean")

    @property
    def context_mode(self) -> DelegationContextMode:
        return DelegationContextMode.ISOLATED

    @property
    def tool_mode(self) -> ToolExecutionMode:
        return ToolExecutionMode.READ

    @property
    def allows_recursive_delegation(self) -> bool:
        return self.allow_recursive_delegation

    def snapshot_mapping(self) -> dict[str, object]:
        return {
            "enabled": True,
            "maxAgentsPerCall": self.max_agents_per_call,
            "maxParallel": self.max_parallel,
            "maxAgentNameChars": self.max_agent_name_chars,
            "maxTitleChars": self.max_title_chars,
            "maxInstructionChars": self.max_instruction_chars,
            "maxObjectiveChars": self.max_objective_chars,
            "maxDepth": self.max_depth,
            "maxAgentsPerRoot": self.max_agents_per_root,
            "contextMode": self.context_mode.value,
            "toolMode": self.tool_mode.value,
            "allowsRecursiveDelegation": self.allows_recursive_delegation,
        }


__all__ = ["DelegationPolicy"]

File: typescript/src/delegation/policy.ts
import { AgentError } from "../shared/errors.js";
import { copyJsonValue } from "../model/validation.js";
import type {
  DelegationDefinition,
  DelegationPolicyOptions,
  DelegationPolicySnapshot,
} from "./types.js";

export class DelegationPolicy {
  readonly #snapshot: DelegationPolicySnapshot;

  public constructor(options: DelegationPolicyOptions = {}) {
    const maxAgentsPerCall = positive(options.maxAgentsPerCall ?? 3, "maxAgentsPerCall");
    const maxParallel = positive(options.maxParallel ?? 3, "maxParallel");
    if (maxParallel > maxAgentsPerCall) {
      throw new TypeError("maxParallel cannot exceed maxAgentsPerCall");
    }
    if (
      options.allowRecursiveDelegation !== undefined
      && typeof options.allowRecursiveDelegation !== "boolean"
    ) {
      throw new TypeError("allowRecursiveDelegation must be boolean");
    }
    this.#snapshot = Object.freeze({
      enabled: true,
      maxAgentsPerCall,
      maxParallel,
      maxAgentNameChars: positive(options.maxAgentNameChars ?? 64, "maxAgentNameChars"),
      maxTitleChars: positive(options.maxTitleChars ?? 120, "maxTitleChars"),
      maxInstructionChars: positive(options.maxInstructionChars ?? 4_000, "maxInstructionChars"),
      maxObjectiveChars: positive(options.maxObjectiveChars ?? 4_000, "maxObjectiveChars"),
      maxDepth: positive(options.maxDepth ?? 3, "maxDepth"),
      maxAgentsPerRoot: positive(options.maxAgentsPerRoot ?? 16, "maxAgentsPerRoot"),
      contextMode: "isolated",
      toolMode: "read",
      allowsRecursiveDelegation: options.allowRecursiveDelegation ?? false,
    });
  }

  public snapshot(): DelegationPolicySnapshot {
    return this.#snapshot;
  }

  public validate(value: unknown): readonly DelegationDefinition[] {
    if (!Array.isArray(value) || value.length < 1 || value.length > this.#snapshot.maxAgentsPerCall) {
      throw new AgentError(
        "invalid_delegation_batch",
        `delegations must contain between one and ${this.#snapshot.maxAgentsPerCall} tasks`,
      );
    }
    const names = new Set<string>();
    return Object.freeze(value.map((item) => {
      if (item === null || typeof item !== "object" || Array.isArray(item)) {
        throw new AgentError("invalid_delegation", "Delegation task must be an object");
      }
      const raw = item as Record<string, unknown>;
      const agentName = bounded(raw.agentName, "agentName", this.#snapshot.maxAgentNameChars);
      if (names.has(agentName)) {
        throw new AgentError("duplicate_delegated_agent", "Delegated Agent names must be unique");
      }
      names.add(agentName);
      const input = raw.input === undefined
        ? Object.freeze({})
        : copyInput(raw.input);
      const priority = raw.priority === undefined ? 0 : integer(raw.priority, "priority");
      if (raw.required !== undefined && typeof raw.required !== "boolean") {
        throw new AgentError("invalid_delegation", "Delegation required must be a boolean");
      }
      return Object.freeze({
        agentName,
        title: bounded(raw.title, "title", this.#snapshot.maxTitleChars),
        instruction: bounded(raw.instruction, "instruction", this.#snapshot.maxInstructionChars),
        objective: bounded(raw.objective, "objective", this.#snapshot.maxObjectiveChars),
        input,
        required: raw.required ?? true,
        priority,
      });
    }));
  }
}

function copyInput(value: unknown): Readonly<Record<string, import("../model/types.js").JsonValue>> {
  const copied = copyJsonValue(value);
  if (copied === null || typeof copied !== "object" || Array.isArray(copied)) {
    throw new AgentError("invalid_delegation", "Delegation input must be an object");
  }
  return copied as Readonly<Record<string, import("../model/types.js").JsonValue>>;
}

function bounded(value: unknown, label: string, maximum: number): string {
  const text = typeof value === "string" ? value.trim() : "";
  if (text === "" || text.length > maximum) {
    throw new AgentError(
      "invalid_delegation",
      `Delegation ${label} must contain between 1 and ${maximum} characters`,
    );
  }
  return text;
}

function positive(value: number, label: string): number {
  if (!Number.isSafeInteger(value) || value < 1) throw new TypeError(`${label} must be positive`);
  return value;
}

function integer(value: unknown, label: string): number {
  if (!Number.isSafeInteger(value)) throw new AgentError("invalid_delegation", `${label} must be an integer`);
  return value as number;
}
